cpp/src/arrow/array/array_dict.cc

```diff
@@ -282,10 +282,11 @@ class DictionaryUnifierImpl : public DictionaryUnifier {
     *out_type = arrow::dictionary(index_type, value_type_);
 
     // Build unified dictionary array
-    std::shared_ptr<ArrayData> data;
-    RETURN_NOT_OK(DictTraits::GetDictionaryArrayData(pool_, value_type_, memo_table_,
-                                                     0 /* start_offset */, &data));
-    *out_dict = MakeArray(data);
+    Result<std::shared_ptr<ArrayData>> data;
+    data = DictTraits::GetDictionaryArrayData(pool_, value_type_, memo_table_,
+                                              0 /* start_offset */);
+    RETURN_NOT_OK(data.status());
+    *out_dict = MakeArray(data.ValueOrDie());
```

Comment:
For `Result`s we have a convenience macro which should save you some boilerplate:
```suggestion
    ARROW_ASSIGN_OR_RAISE(auto data, DictTraits::GetDictionaryArrayData(
        pool_, value_type_, memo_table_, /*start_offset=*/0));
    *out_dict = MakeArray(data);
```

Reply:
Awesome! I was trying to find such a macro and didn't succeed. Thanks for the pointer!

Reply:
Done.